### ISSUE TITLE
AML: Implement CondRefOf

### ISSUE THREAD

**Issue by rw-vanc** [rust-osdev/acpi#141]
On HP-dq3xxx, the following asl/aml occurs:
`If (CondRefOf (PU2C))`
`5B 12 50 55 32 43 00`

**Comment by rw-vanc** [comment]
Which file should cond_ref_of be implemented in? ref_of is also not implemented.

**Comment by IsaacWoods** [comment]
Yep, again not supported yet. The opcodes aren't handled mainly because the machinery for refs isn't there yet - but shouldn't be too hard to build out.

They are expression opcodes, so should be in `expression.rs`.

**Comment by rw-vanc** [comment]
This is an example from the [uefi spec](https://uefi.org/htmlspecs/ACPI_Spec_6_4_html/19_ASL_Reference/ACPI_Source_Language_Reference.html#index-with-packages)
```
Store (DeRefOf (Index (DeRefOf (Index (IO0D, 0)), 5)), Local0)
```
What do suggest for the representation of these things in the AML parser?

**Comment by IsaacWoods** [comment]
> `Store (DeRefOf (Index (DeRefOf (Index (IO0D, 0)), 5)), Local0)`

Hm yeah this is about as complex as these expressions will get afaik, but ofc we'll need to support them fully eventually so might as well now. Basically, I think we need to introduce a new value type (`Reference`) that stores *how* to access a value in the namespace (e.g. it's the 5th thing in `IO0D`, where that object can be a buffer, string, or package). Of course, this representation will need to be dynamic (i.e. probably end up as some boxed enums or something) to deal with arbitrary cases like your example.

I was originally dreading dealing with these because I thought that references could appear in all sorts of places (i.e. instead of regular objects, and you'd have to deal with implicitly dereferencing them), but it doesn't actually look that way! I *think* AML requires you to manually use `DeRefOf` when you want to "resolve" a reference into a real object, and so this should be handled by taking the given reference and running the resolution logic (e.g. look up the object from the namespace (or a local or whatever), and doing the actual index into it to get the current value of the referenced object).

`CondRefOf` is slightly more complex in that it needs to gracefully handle the referenced object not actually existing, but that shouldn't be too much worse.

So overall this doesn't actually sound too bad. Some existing types (maybe `Target`) might be useful. This is just my thoughts in the last 10 mins of thinking and reading the spec so don't take it as scripture, but hopefully that's helpful, and feel free to reach out with any questions and I'll get to them when I can.
